<a href="https://colab.research.google.com/github/Rkitenge91/MATH5010-repository/blob/main/Raissa_Kitenge_Exam3_MATH5010.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#MATH5010 - EXAM 3 - RAISSA KITENGE

In [4]:
import numpy as np
from math import erf, sqrt
import mpmath as mp

# Question 3 — CONCRETE DATA COMPUTATION FROM QUESTION 2 FORMULAS.

For each model below, you observe a small sample. Compute:

- The Maximum Likelihood Estimator (MLE)
- The Method of Moments (MOM) estimator
- The posterior parameter distribution under the given conjugate prior
- The Bayes estimator (posterior mean)
- The MAP estimator


## 3a. Bernoulli(p)

### Data
We observe:

$$
x = (1,0,1,1,0,1,1,1,0,1)
$$

There are:

$$
n = 10, \quad S = \sum_{i=1}^n x_i
$$


### Prior
We assume a Beta prior:

$$
p \sim \text{Beta}(\alpha = 2,\; \beta = 2)
$$

---

### Formulas

**1. MLE**

$$
\hat{p}_{MLE} = \frac{S}{n}
$$

**2. MOM estimator**

$$
\hat{p}_{MOM} = \frac{S}{n}
$$

**3. Posterior distribution**

$$
p \mid x \sim \text{Beta}(\alpha + S,\; \beta + n - S)
$$

**4. Posterior mean (Bayes estimator)**

$$
E[p \mid x] = \frac{\alpha + S}{\alpha + \beta + n}
$$

**5. MAP estimator**

$$
\hat{p}_{MAP} =
\frac{(\alpha + S) - 1}{(\alpha + \beta + n) - 2}
$$

In [14]:
x = np.array([1,0,1,1,0,1,1,1,0,1])
n = len(x)
S = x.sum()

mle = S / n
mom = S / n

alpha_post = 2 + S
beta_post = 2 + n - S

bayes = alpha_post / (alpha_post + beta_post)

map_est = (alpha_post - 1) / (alpha_post + beta_post - 2)

mle, mom, alpha_post, beta_post, bayes, map_est

(np.float64(0.7),
 np.float64(0.7),
 np.int64(9),
 np.int64(5),
 np.float64(0.6428571428571429),
 np.float64(0.6666666666666666))

## 3b. Normal(μ, σ² = 4)

### Data
We observe:

$$
x = (3,5,4,6)
$$

Sample mean:

$$
\bar{x} = \frac{1}{4}\sum_{i=1}^4 x_i
$$

Known variance:

$$
\sigma^2 = 4
$$

---

### Prior

$$
\mu \sim N(\mu_0 = 0,\ \tau^2 = 4)
$$

---

### Formulas

**1. MLE**

$$
\hat{\mu}_{MLE} = \bar{x}
$$

**2. MOM estimator**

$$
\hat{\mu}_{MOM} = \bar{x}
$$

**3. Posterior mean**

$$
\mu' =
\frac{\tau^2 n \bar{x} + \sigma^2 \mu_0}{n\tau^2 + \sigma^2}
$$

**4. Posterior variance**

$$
(\tau')^2 =
\frac{\sigma^2 \tau^2}{n\tau^2 + \sigma^2}
$$

**5. MAP estimator**

Since the posterior is Normal:

$$
\hat{\mu}_{MAP} = \mu'
$$

In [15]:
x = np.array([3,5,4,6])
n = len(x)
sigma2 = 4
tau2 = 4
mu0 = 0

xbar = x.mean()

mle = xbar
mom = xbar

post_mean = (tau2*n*xbar + sigma2*mu0) / (n*tau2 + sigma2)
post_var = (sigma2 * tau2) / (n*tau2 + sigma2)

map_est = post_mean

mle, mom, post_mean, post_var, map_est

(np.float64(4.5), np.float64(4.5), np.float64(3.6), 0.8, np.float64(3.6))

## 3c. Poisson(λ)

###Data:
$$
x = (2,3,4,1,0), \qquad n = 5
$$

###Sum and sample mean:
$$
S = \sum_{i=1}^5 x_i = 2 + 3 + 4 + 1 + 0 = 10
$$
$$
\bar{x} = \frac{S}{n} = \frac{10}{5} = 2
$$

###Prior:
$$
\lambda \sim \text{Gamma}(\alpha = 3,\; \beta = 1)
$$

### MLE:
$$
\hat{\lambda}_{MLE} = \bar{x} = 2
$$

### MOM estimator:
For Poisson, the mean equals λ:
$$
\hat{\lambda}_{MOM} = \bar{x} = 2
$$

### Posterior distribution:
For the Poisson–Gamma conjugate model:
$$
\lambda \mid x \sim \text{Gamma}(\alpha + S,\; \beta + n)
$$

Here:
$$
\alpha + S = 3 + 10 = 13,
\qquad
\beta + n = 1 + 5 = 6
$$

So the posterior is:
$$
\lambda \mid x \sim \text{Gamma}(13, 6)
$$

### Bayes estimator (posterior mean):
$$
E[\lambda \mid x] = \frac{\alpha + S}{\beta + n}
= \frac{13}{6} \approx 2.1667
$$

### MAP estimator:
$$
\hat{\lambda}_{MAP}
= \frac{(\alpha + S) - 1}{\beta + n}
= \frac{13 - 1}{6}
= 2
$$

In [17]:
x = np.array([2, 3, 4, 1, 0])
n = len(x)

S = x.sum()
mle = x.mean()
mom = mle

alpha_prior, beta_prior = 3, 1

alpha_post = alpha_prior + S
beta_post = beta_prior + n

bayes = alpha_post / beta_post
map_est = (alpha_post - 1) / beta_post

S, mle, mom, alpha_post, beta_post, bayes, map_est

(np.int64(10),
 np.float64(2.0),
 np.float64(2.0),
 np.int64(13),
 6,
 np.float64(2.1666666666666665),
 np.float64(2.0))

## 3d. Exponential(λ) with density

$$
f(x \mid \lambda) = \lambda e^{-\lambda x}
$$

###Data:
$$
x = (1,2,3,1,3), \qquad n = 5
$$

###Sum and sample mean:
$$
T = \sum_{i=1}^5 x_i = 1 + 2 + 3 + 1 + 3 = 10
$$
$$
\bar{x} = \frac{T}{n} = \frac{10}{5} = 2
$$

###Prior:
$$
\lambda \sim \text{Gamma}(\alpha = 2,\; \beta = 1)
$$

### MLE:
$$
\hat{\lambda}_{MLE} = \frac{n}{T} = \frac{5}{10} = 0.5
$$

### MOM estimator:
Since the exponential mean is \(1/\lambda\):
$$
\hat{\lambda}_{MOM} = \frac{1}{\bar{x}} = \frac{1}{2} = 0.5
$$

### Posterior distribution:
For exponential + Gamma(α, β) conjugacy:
$$
\lambda \mid x \sim \text{Gamma}(\alpha + n,\; \beta + T)
$$

Compute:
$$
\alpha + n = 2 + 5 = 7,
\qquad
\beta + T = 1 + 10 = 11
$$

So the posterior is:
$$
\lambda \mid x \sim \text{Gamma}(7, 11)
$$

### Bayes estimator (posterior mean):
$$
E[\lambda \mid x] = \frac{\alpha + n}{\beta + T}
= \frac{7}{11}
\approx 0.6364
$$

### MAP estimator:
$$
\hat{\lambda}_{MAP}
= \frac{(\alpha + n) - 1}{\beta + T}
= \frac{7 - 1}{11}
= \frac{6}{11}
\approx 0.5455
$$

In [19]:
x = np.array([1, 2, 3, 1, 3])
n = len(x)

T = x.sum()

mle = n / T
mom = 1 / x.mean()

alpha_prior, beta_prior = 2, 1

alpha_post = alpha_prior + n
beta_post = beta_prior + T

bayes = alpha_post / beta_post
map_est = (alpha_post - 1) / beta_post

T, mle, mom, alpha_post, beta_post, bayes, map_est

(np.int64(10),
 np.float64(0.5),
 np.float64(0.5),
 7,
 np.int64(11),
 np.float64(0.6363636363636364),
 np.float64(0.5454545454545454))

# Question 4 — Bernoulli Confidence Intervals and Tests with Bayesian Comparison.
A startup is testing a new landing page. During one week, they record whether each of n = 50
visitors clicks the "Sign Up" button. Out of these 50 visitors, x = 18 clicked.
We model each visitor's behavior as i.i.d. Bernoulli(p), where p is the (unknown) click-through probability.

We observe:
- n = 50 visitors
- x = 18 clicks

### 4a. Compute the Sample Proportion $\hat{p}$

The sample proportion is defined as:

$$
\hat{p} = \frac{x}{n}
$$

In [8]:
## 4a. Sample Proportion
p_hat = 18/50
p_hat

0.36

#4b. Construct an approximate 95% confidence interval for p using the normal approximation to the sampling distribution $\hat{p}$. Clearly write the formula and plug in the numbers

We use the normal approximation to the sampling distribution of $\hat{p}$.

**General formula:**

$$
\hat{p} \;\pm\; z_{0.975} \sqrt{ \frac{\hat{p}(1 - \hat{p})}{n} }
$$

where the 97.5th percentile of the standard normal distribution is

$$
z_{0.975} = 1.96.
$$

---

### Plug in the values:

Given:

$$
\hat{p} = 0.36, \quad n = 50,
$$

the standard error is:

$$
\sqrt{ \frac{0.36(1 - 0.36)}{50} }.
$$

Thus, the 95% confidence interval is:

$$
0.36 \;\pm\; 1.96 \sqrt{ \frac{0.36(1 - 0.36)}{50} }.
$$

In [20]:
import numpy as np

n = 50
z = 1.96

se = np.sqrt(p_hat*(1-p_hat)/n)
ci_lower = p_hat - z*se
ci_upper = p_hat + z*se

ci_lower, ci_upper

(np.float64(0.22695078805193922), np.float64(0.49304921194806073))

## 4c. Hypothesis Test for H0: p = 0.5 vs H1: p ≠ 0.5

We use a large-sample z-test for the proportion.


### **(i) Test Statistic Z**

The test statistic is

$$
Z = \frac{\hat{p} - p_0}{\sqrt{ \frac{p_0(1 - p_0)}{n} }}
$$

with $${ p_0 = 0.5 }.$$


### **(ii) Approximate Two-Sided p-value**

For a two-sided test:

$$
p\text{-value} = 2\bigl(1 - \Phi(|Z|)\bigr),
$$

where \$$\Phi $$is the CDF of the standard normal distribution.

### **(iii) Conclusion in context**

At significance level α = 0.05:

- If the p-value < 0.05, we reject H₀ and conclude that the true click-through rate p is different from 0.5.
- If the p-value ≥ 0.05, we fail to reject H₀ and conclude that the data do not provide strong evidence that p differs from 0.5.

In [21]:
# Question 4c: hypothesis test for H0: p = 0.5 vs H1: p != 0.5

n = 50
x = 18
p_hat = x / n
p0 = 0.5

# (i) Test statistic Z
se0 = np.sqrt(p0 * (1 - p0) / n)
Z = (p_hat - p0) / se0

# (ii) Approximate two-sided p-value
Phi = lambda z: 0.5 * (1 + erf(z / np.sqrt(2)))
p_value = 2 * (1 - Phi(abs(Z)))

print("i) Test statistic Z =", Z)

print("ii) Two-sided p-value =", p_value)

# (iii) Conclusion in context at alpha = 0.05
alpha = 0.05
if p_value < alpha:
    print("iii) Conclusion: Reject H0 at the 5% level.")
    print("    The data provide evidence that the true click-through rate p is different from 0.5")
    print("    (and since p̂ = {:.2f} < 0.50, it appears lower than 0.5).".format(p_hat))
else:
    print("iii) Conclusion: Fail to reject H0 at the 5% level.")
    print("    The data do not provide strong evidence that the true click-through rate p differs from 0.5.")

i) Test statistic Z = -1.9798989873223334
ii) Two-sided p-value = 0.04771488023735104
iii) Conclusion: Reject H0 at the 5% level.
    The data provide evidence that the true click-through rate p is different from 0.5
    (and since p̂ = 0.36 < 0.50, it appears lower than 0.5).


### 4d. Bayesian analysis with Beta(1,1) prior

We assume a prior
$$
p \sim \text{Beta}(1,1).
$$

This is the same as a **Uniform(0,1)** prior (all values of \(p\) between 0 and 1 are equally likely before seeing the data).

We observed:
$$
n = 50, \quad x = 18.
$$


#### 1. Posterior distribution of \(p\)

For a Bernoulli model with prior $p \sim \text{Beta}(\alpha,\beta)$ and data with $x$ successes out of $n$ trials, the posterior is

$$
p \mid x \sim \text{Beta}(\alpha + x,\; \beta + n - x).
$$

Here $\alpha = 1$, $\beta = 1$, $x = 18$, and $n = 50$, so

$$
\alpha + x = 1 + 18 = 19,
\qquad
\beta + n - x = 1 + 50 - 18 = 33.
$$

Therefore,
$$
p \mid \text{data} \sim \text{Beta}(19, 33).
$$


#### 2. Posterior mean (Bayes estimator)

For a Beta\((\alpha,\beta)\) distribution,

$$
E[p] = \frac{\alpha}{\alpha + \beta}.
$$

So the posterior mean here is

$$
E[p \mid \text{data}] =
\frac{19}{19 + 33}
= \frac{19}{52}.
$$


#### 3. 95% central Bayesian credible interval

A 95% central credible interval uses the 2.5% and 97.5% quantiles of the posterior distribution.

Since the posterior distribution is Beta(19, 33), the 95% central credible interval is:

$$
\big(
\text{Beta}^{-1}(0.025; 19, 33),\;
\text{Beta}^{-1}(0.975; 19, 33)
\big)
$$

Here “Beta⁻¹(q; 19, 33)” simply means the **q-th quantile** of a Beta(19, 33) distribution.

In [24]:
# Question 4d: posterior for p with Beta(1,1) prior

n = 50
x = 18

alpha_prior = 1
beta_prior = 1

# Posterior parameters
alpha_post = alpha_prior + x
beta_post = beta_prior + (n - x)

# Posterior mean
post_mean = alpha_post / (alpha_post + beta_post)

print("Posterior distribution: p | data ~ Beta({}, {})".format(alpha_post, beta_post))
print("Posterior mean E[p | data] =", post_mean)

Posterior distribution: p | data ~ Beta(19, 33)
Posterior mean E[p | data] = 0.36538461538461536


## 4e. Bayesian Decision for H1: p < 0.5

Compute the posterior probability:

$$
P(p < 0.5 \mid x)
= F_{\text{Beta}(19,33)}(0.5)
$$

Reject H0 at α = 0.05 if:

$$
P(p > 0.5 \mid x) < 0.05.
$$

### Bayesian decision rule at the 5% level

In the Bayesian framework:

- If the posterior probability of the alternative hypothesis is greater than 95%,
  we decide in favor of \( H_1 \).
  
Thus, we compare:

$$
P(p < 0.5 \mid \text{data}) \quad \text{to} \quad 0.95.
$$

If
$$
P(p < 0.5 \mid \text{data}) > 0.95,
$$
we conclude that \( p < 0.5 \).

Otherwise, we do **not** conclude that \( p < 0.5 \).

We compute this probability numerically using the Beta CDF.

In [26]:
# Question 4e: posterior probability P(p < 0.5 | data)
# Posterior from part (d): Beta(19, 33)

alpha_post = 19
beta_post = 33

# Beta CDF
posterior_prob = mp.betainc(alpha_post, beta_post, 0, 0.5, regularized=True)

print("Posterior probability P(p < 0.5 | data) =", posterior_prob)

Posterior probability P(p < 0.5 | data) = 0.975563053840865


### Conclusion

From the posterior distribution p | data ~ Beta(19, 33), we compute

P(p < 0.5 | data) ≈ 0.9755.

At the 5% decision level, we compare this value to 0.95:

- If the posterior probability is greater than 0.95, we conclude that p < 0.5.
- If the posterior probability is not greater than 0.95, we do not conclude that p < 0.5.

Since 0.9755 > 0.95, the posterior evidence is strong enough to conclude that the true click-through rate p is less than 0.5.

**Final decision:**  
We conclude that p < 0.5 in this Bayesian framework.

# Question 5 — One-Way ANOVA for Fertilizer Yields

We test the effect of three fertilizers (A, B, C) on crop yields.  

The observed yields are:

$$
\begin{aligned}
\text{Fertilizer A:} &\quad 20,\; 22,\; 19,\; 21 \\
\text{Fertilizer B:} &\quad 25,\; 27,\; 26,\; 28 \\
\text{Fertilizer C:} &\quad 23,\; 24,\; 22,\; 21
\end{aligned}
$$

There are \( 3 \) groups (treatments) and \( 4 \) observations per group, for a total of \( 12 \) observations.

## 5a.State the one-way ANOVA model and the null and alternative hypotheses about the treatment means.
One-way ANOVA model:

$$
Y_{ij} = \mu + \tau_i + \epsilon_{ij}
$$

with:
$$
\epsilon_{ij} \sim N(0, \sigma^2)
$$
where:
- $\mu$: overall mean  
- $\tau_i$: effect of treatment $i$  
- $\epsilon_{ij} \sim N(0,\sigma^2)$: random error
- $Y_{ij}$ is observation $j$ in treatment group $i$  


**Null hypothesis:**

$$
H_0: \mu_1 = \mu_2 = \mu_3
$$

**Alternative hypothesis:**

At least one group mean is different.

## 5b. Compute the Sample Means for each group and the overall Mean

Group means:

$$
\bar{Y}_A = \frac{20 + 22 + 19 + 21}{4}
$$

$$
\bar{Y}_B = \frac{25 + 27 + 26 + 28}{4}
$$

$$
\bar{Y}_C = \frac{23 + 24 + 22 + 21}{4}
$$

The overall mean is:

$$
\bar{Y} = \frac{1}{12} \sum_{i=1}^{12} Y_i
$$

In [27]:
A = np.array([20,22,19,21])
B = np.array([25,27,26,28])
C = np.array([23,24,22,21])

means = [A.mean(), B.mean(), C.mean()]
overall = np.concatenate([A,B,C]).mean()

means, overall

([np.float64(20.5), np.float64(26.5), np.float64(22.5)],
 np.float64(23.166666666666668))

## 5c. Compute the ANOVA Sums of Squares

Between treatments:

$$
SS_{Tr} = \sum_{i=1}^3 n_i(\bar{Y}_i - \bar{Y})^2
$$

Error:

$$
SS_E = \sum_{i=1}^3 \sum_{j=1}^4 (Y_{ij} - \bar{Y}_i)^2
$$

Total:

$$
SS_{Tot} = \sum_{i=1}^{12} (Y_i - \bar{Y})^2
$$

In [28]:
groups = [A, B, C]
n_i = 4

SS_Tr = sum(n_i*(g.mean() - overall)**2 for g in groups)
SS_E = sum(((g - g.mean())**2).sum() for g in groups)

all_data = np.concatenate(groups)
SS_Tot = ((all_data - overall)**2).sum()

SS_Tr, SS_E, SS_Tot

(np.float64(74.66666666666667),
 np.float64(15.0),
 np.float64(89.66666666666666))

## 5d. Construct the ANOVA Table: List degrees of freedom, mean squares and F Statistic

### Step 1: Sums of Squares (from part 5c)

- SS_Tr (Between Treatments) = 74.6667  
- SS_E (Error/Within) = 15  
- SS_Total = 89.6667  

### Step 2: Degrees of Freedom

- df_Tr = k − 1 = 3 − 1 = 2  
- df_E = n − k = 12 − 3 = 9  
- df_Total = n − 1 = 12 − 1 = 11  

### Step 3: Mean Squares

Mean squares are computed by dividing SS by the corresponding df:

- MS_Tr = SS_Tr / df_Tr  
  = 74.6667 / 2  
  = 37.3333  

- MS_E = SS_E / df_E  
  = 15 / 9  
  = 1.6667  

### Step 4: F Statistic

F = MS_Tr / MS_E  
  = 37.3333 / 1.6667  
  ≈ 22.4  


In [29]:
df_tr = 2
df_e = 9

MS_Tr = SS_Tr / df_tr
MS_E = SS_E / df_e

F_stat = MS_Tr / MS_E

MS_Tr, MS_E, F_stat

(np.float64(37.333333333333336),
 np.float64(1.6666666666666667),
 np.float64(22.400000000000002))

## ANOVA Table

| Source                | SS        | df  | MS        | F        |
|-----------------------|-----------|-----|-----------|----------|
| Treatments (between)  | 74.6667   | 2   | 37.3333   | 22.4     |
| Error (within)        | 15        | 9   | 1.6667    |          |
| Total                 | 89.6667   | 11  |           |          |


## 5e. Decision Rule (α = 0.05)

Reject the null hypothesis if:

$$
F > F_{2,9;0.95}
$$

Otherwise, fail to reject.


Our computed F statistic is:

F_stat = 22.4

To make a decision at the α = 0.05 significance level, we compare this value to the critical value from the F distribution:

F_critical = F(2, 9, 0.95)

Since the observed F statistic (22.4) is much larger than the critical value, we reject the null hypothesis.

**Conclusion:**  
There is strong evidence that at least one of the fertilizer treatments has a different mean crop yield compared to the others. Hence, we reject the null hypothesis.